### PR TITLE
YJIT: Use a boxed slice for gc_obj_offsets

### PR DESCRIPTION
**Pull request by k0kubun**
This reduces `yjit_alloc_size` by 0.15MB (1.7%) on 25 itrs of railsbench with arm64-darwin22 stats build. Not a huge impact, but this field is immutable and it seems like a simple and straightforward improvement.

## Before
```
yjit_alloc_size:           9,045,001
```

## After
```
yjit_alloc_size:           8,894,479
```

### CODE REVIEW

File: yjit/src/stats.rs [hunk 1]
```diff
@@ -141,6 +141,19 @@ macro_rules! make_counters {
...
+    // once before any Ruby code executes
+    ($counter_name:ident, $count:expr) => {
+        #[allow(unused_unsafe)]
+        {
+            unsafe { $crate::stats::COUNTERS.$counter_name += $count as u64 }
+        }
+    };
+}
+pub(crate) use add_counter;
+
```
Comment (XrXr) [inline yjit/src/stats.rs:156]:
```suggestion
```
seems like this was added by accident?

**Review by XrXr — COMMENTED** [review]
I think we should use get rid of the `Option` and just have `Box<[u32]>`. They have the same layout (boxed slices are "fat", 16 byte pointers).
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7e2c884197eb871080c58ae47ff2fb7f

File: yjit/src/stats.rs (hunk 1 above)
Comment (k0kubun) [reply yjit/src/stats.rs:156]: What do you mean? Do we have other ways to increase a counter by a non-1 amount?

Comment (XrXr) [reply yjit/src/stats.rs:156]: Oh sorry, I failed to see the call site.

**Comment by k0kubun** [comment]
> I think for simplicity we should use get rid of the Option and just have Box<[u32]>. They have the same layout (boxed slices are "fat", 16 byte pointers).
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7e2c884197eb871080c58ae47ff2fb7f

Nice, thank you :+1: Fixed it in https://github.com/ruby/ruby/pull/7397/commits/5a5eada3b1e35f9bfa6ed1764bc9cac87055fd33.

**Review by XrXr — APPROVED** [review]

File: yjit/src/stats.rs [hunk 2]
```diff
@@ -141,6 +141,19 @@ macro_rules! make_counters {
     }
 }
 
+/// Macro to increase a counter by name and count
+macro_rules! add_counter {
```
Comment (maximecb) [inline yjit/src/stats.rs:145]:
```suggestion
macro_rules! counter_incr_by {
```

Or, `counter_add_to`

`add_counter` sounds like you're adding a new counter.

Comment (k0kubun) [reply yjit/src/stats.rs]: Agreed, and your version sounds better. Fixed in https://github.com/ruby/ruby/pull/7397/commits/6e9e9cf8c27f81d0647942f4d011b9deff51434a.

**Review by maximecb — APPROVED** [review]

**Pull request merged**